### PR TITLE
RecordData Relationship Tests

### PR DESCRIPTION
**Pull request by igorT**
Add tests for custom RecordData relationship handling.

### CODE REVIEW

File: packages/-ember-data/tests/helpers/public-props.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+export default function publicArgs(publicArgs, obj) {
```
Comment (hjdivad) [inline packages/-ember-data/tests/helpers/public-props.js]: minor: function name should match the file name (`publicArgs` vs `public-props`)

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts [hunk 2]
```diff
@@ -323,4 +367,294 @@ module('integration/record-data - Custom RecordData Implementations', function (
...
+        super();
+        this._storeWrapper = storeWrapper;
+      }
+
+      getBelongsTo(key: string) {
+        assert.equal(key, 'landlord', 'Passed correct key to getBelongsTo');
+        return belongsToReturnValue;
+      }
+
+      setDirtyBelongsTo(key: string, recordData: this | null) {
```
Comment (hjdivad) [inline packages/-ember-data/tests/integration/record-data/record-data-test.ts]: did we change from `setBelongsTo` to `setDirtyBelongsTo`?

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts [hunk 3]
```diff
@@ -323,4 +367,294 @@ module('integration/record-data - Custom RecordData Implementations', function (
...
+      constructor(storeWrapper) {
+        super();
+        this._storeWrapper = storeWrapper;
+      }
+
+      getHasMany(key: string) {
+        return hasManyReturnValue;
+      }
+
+      // TODO: investigate addToHasMany being called during unloading
```
Comment (hjdivad) [inline packages/-ember-data/tests/integration/record-data/record-data-test.ts]: as a separate item, let's add this as a skipped test to make it less likely to get lost

File: packages/-ember-data/tests/integration/record-data/store-wrapper-test.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  landlord;
+
+  @belongsTo('car', { async: false })
+  car;
+
+  @hasMany('person', { async: false })
+  tenants;
+}
+
+// TODO Make the class implement the RD interface
```
Comment (hjdivad) [inline packages/-ember-data/tests/integration/record-data/store-wrapper-test.ts]:
```
TODO: this should work
class TestRecordData implements RecordData
```

File: packages/-ember-data/tests/integration/record-data/store-wrapper-test.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  hooks.beforeEach(function () {
+    let { owner } = this;
+
+    owner.register('model:person', Person);
+    owner.register('model:house', House);
+    owner.register('model:car', Car);
+    owner.register('service:store', CustomStore);
+  });
+
+  test("Relationship definitions", async function (assert) {
```
Comment (hjdivad) [inline packages/-ember-data/tests/integration/record-data/store-wrapper-test.ts]: this test looks good; we should probably add tests for the rest of the store wrapper api though no?

Comment (igorT) [reply packages/-ember-data/tests/integration/record-data/store-wrapper-test.ts]: added

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 2 above)
Comment (igorT) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: Not sure, will investigate

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 3 above)
Comment (igorT) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: 👍

File: packages/-ember-data/tests/helpers/public-props.js (hunk 1 above)
Comment (igorT) [reply packages/-ember-data/tests/helpers/public-props.js]: fixed

File: packages/-ember-data/tests/integration/record-data/store-wrapper-test.ts (hunk 4 above)
Comment (igorT) [reply packages/-ember-data/tests/integration/record-data/store-wrapper-test.ts]: good point, fixed

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 2 above)
Comment (runspired) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: We did

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts [hunk 6]
```diff
@@ -12,7 +12,52 @@ class Person extends Model {
...
+
+const runspiredHash = {
+  type: 'person',
+  id: '2',
+  attributes: {
+    name: 'Runspired'
+  }
+};
+
+const igorHash = {
```
Comment (runspired) [inline packages/-ember-data/tests/integration/record-data/record-data-test.ts]: defining these outside of the context of a `beforeEach` is leaky.

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts [hunk 7]
```diff
@@ -323,4 +367,294 @@ module('integration/record-data - Custom RecordData Implementations', function (
...
+        this._storeWrapper = storeWrapper;
+      }
+
+      getBelongsTo(key: string) {
+        assert.equal(key, 'landlord', 'Passed correct key to getBelongsTo');
+        return belongsToReturnValue;
+      }
+
+      setDirtyBelongsTo(key: string, recordData: this | null) {
+        belongsToReturnValue = { data: { id: '3', type: 'person' } };
```
Comment (runspired) [inline packages/-ember-data/tests/integration/record-data/record-data-test.ts]: we should add a note that this is an intentional misdirection

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts [hunk 8]
```diff
@@ -323,4 +367,294 @@ module('integration/record-data - Custom RecordData Implementations', function (
...
+      data: [houseHash]
+    });
+
+    let house = store.peekRecord('house', '1');
+    assert.equal(house.get('landlord.name'), 'David', 'belongsTo get correctly looked up');
+
+    let runspired = store.peekRecord('person', '2');
+    house.set('landlord', runspired);
+
+    assert.equal(house.get('landlord.name'), 'Igor', 'RecordData sets the custom belongsTo value');
```
Comment (runspired) [inline packages/-ember-data/tests/integration/record-data/record-data-test.ts]: also note the intentional misdirection here

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 3 above)
Comment (runspired) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: it's probably related to treating unloading of sync relationships as client-side-deletes

**Comment by igorT** [comment]
@runspired yeah, I was thinking about it and iterated with @hjdivad to get them to be understandbale. Adding comments seems :+1: though

**Comment by igorT** [comment]
The main reason they are misdirectional is to be certain that things don't accidentally work because the rest of our stack does a reasonable thing

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 6 above)
Comment (igorT) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: fixed

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 7 above)
Comment (igorT) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: fixed at call site

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 8 above)
Comment (igorT) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: fixed at call site

File: packages/-ember-data/tests/integration/record-data/record-data-test.ts (hunk 2 above)
Comment (hjdivad) [reply packages/-ember-data/tests/integration/record-data/record-data-test.ts]: seems bad?  this is public api?  what was the reason to make this change?

**Pull request merged**